README.md
# lalr-parser-test
## Testing how different LALR(1) parsers detect grammar conflicts

The `lemon` parser here have several modifications like:
- Command line option to convert from lemon grammar to yacc grammar `-y           Print yacc grammar without actions.`
- Command line option to convert from lemon grammar to yacc grammar with full precedences `-Y           Print yacc grammar without actions with full precedences.`
- Command line option to use yacc rule precedences `-z           Use yacc rule precedence`
- Command line option to ignore all precedences `-u           Ignore all precedences`
- Added a new directive "%yacc_prec" to use yacc rule precedences (the same as command line `-u`).
- Enhanced command line option to output a naked lemon grammar `-g           Print grammar without actions.`

The byacc parser (from https://invisible-island.net/byacc/byacc.html) here have several modifications like:
- Command line option to print naked yacc `-n                            generate naked grammar`
- Command line option to convert from yacc grammar to lemon grammar `-E                    write lemon grammar`
- Command line option to convert from yacc grammar to EBNF grammar (understood by https://www.bottlecaps.de/rr/ui) `-e                    write ebnf grammar`
- Command line option to ignore all precedences `-u                    ignore precedences`
- Command line option to use lemon rule precedences `-z                    use leftmost token for rule precedence`
- Add code to accept/skip bison styles alias for non-terminals `rule[alias]` 

The bison-3.8.2 parser here (only the diff) have several modifications like:
- Command line option to print naked yacc `-n                            generate naked grammar`
- Command line option to convert from yacc grammar to lemon grammar `-N                            generate lemon grammar`
- Command line option to convert from yacc grammar to EBNF grammar (understood by https://www.bottlecaps.de/rr/ui) `-e                    write ebnf grammar`
- Command line option to ignore all precedences `-z                    ignore precedences`


All of the above was made to make easier to compare how lemon/byacc/bison parse LARL(1) grammars.

Obs: On the examples showun bellow `lemon-nb` and `byacc-nb` stands for the lemon/byacc parser from this repository,
`mybison` stands for bison-3.8.2 modified with an extra command line `-z, --noprec                     ignore all precedences`.

## Testing sqlite3 grammar:
```
$>lemon-nb -Y sqlite3.yl > sqlite3.y #convert lemon grammar to yacc grammar

$>byacc-nb -E sqlite3.y # convert yacc grammar to lemon grammar
byacc-nb: 53 reduce/reduce conflicts.
53 conflicts
185 terminal symbols
134 non-terminal symbols
319 total symbols
445 rules
887 states

$>mv y.yl sqlite3-y.yl # rename converted lemon grammar

$>lemon-nb -s sqlite3-y.yl
Parser statistics:
  terminal symbols...................   184
  non-terminal symbols...............   133
  total symbols......................   317
  rules..............................   442
  states.............................   602
  conflicts..........................     0
  action table entries...............  2078
  lookahead table entries............  2262
  total table size (bytes)........... 11956

$>lemon-nb -s -u sqlite3-y.yl # parse again but ignoring all precedences
827 parsing conflicts.
Parser statistics:
  terminal symbols...................   184
  non-terminal symbols...............   133
  total symbols......................   317
  rules..............................   442
  states.............................   602
  conflicts..........................   827
  conflicts S/R......................   788
  conflicts R/R......................    39
  action table entries...............  2000
  lookahead table entries............  2184
  total table size (bytes)........... 11644
 
$>byacc-nb  sqlite3.y
byacc-nb: 53 reduce/reduce conflicts.
53 conflicts
185 terminal symbols
134 non-terminal symbols
319 total symbols
445 rules
887 states

$>byacc-nb -u sqlite3.y # parse again but ignoring all precedences
byacc-nb: 788 shift/reduce conflicts, 39 reduce/reduce conflicts.
827 conflicts
185 terminal symbols
134 non-terminal symbols
319 total symbols
445 rules
887 states

$>mybison  sqlite3.y
sqlite3.y: warning: 53 reduce/reduce conflicts [-Wconflicts-rr]
sqlite3.y: note: rerun with option '-Wcounterexamples' to generate conflict counterexamples

$>mybison -z sqlite3.y # parse again but ignoring all precedences
sqlite3.y: warning: 734 shift/reduce conflicts [-Wconflicts-sr]
sqlite3.y: warning: 93 reduce/reduce conflicts [-Wconflicts-rr]
sqlite3.y: note: rerun with option '-Wcounterexamples' to generate conflict counterexamples
```

## Testing cfront3 grammar (https://github.com/mingodad/cfront-3/blob/master/src/gram.y):
```
$>byacc-nb -E -n cfront3.y # convert yacc grammar to lemon grammar, also generate naked yacc grammar
byacc-nb: 21 shift/reduce conflicts, 3 reduce/reduce conflicts.
24 conflicts
95 terminal symbols
115 non-terminal symbols
210 total symbols
417 rules
709 states

$>mv y.yl cfront3.yl # rename converted lemon grammar

$>mv y.yn cfront3-naked.y # rename naked yacc grammar

$>lemon-nb -s cfront3.yl
19 parsing conflicts.
Parser statistics:
  terminal symbols...................    94
  non-terminal symbols...............    92
  total symbols......................   186
  rules..............................   391
  states.............................   421
  conflicts..........................    19
  conflicts S/R......................    19
  conflicts R/R......................     0
  action table entries...............  5360
  lookahead table entries............  5360
  total table size (bytes)........... 18272

$>lemon-nb -s -u cfront3.yl # parse again but ignoring all precedences
cfront3.yl:230: This rule can not be reduced.

1077 parsing conflicts.
Parser statistics:
  terminal symbols...................    94
  non-terminal symbols...............    92
  total symbols......................   186
  rules..............................   391
  states.............................   423
  conflicts..........................  1077
  conflicts S/R......................  1074
  conflicts R/R......................     3
  action table entries...............  5167
  lookahead table entries............  5202
  total table size (bytes)........... 17736

$>byacc-nb  cfront3-naked.y
byacc-nb: 21 shift/reduce conflicts, 3 reduce/reduce conflicts.
24 conflicts
95 terminal symbols
92 non-terminal symbols
187 total symbols
394 rules
686 states

$>byacc-nb -u cfront3-naked.y # parse again but ignoring all precedences
byacc-nb: 1 rule never reduced
byacc-nb: 1074 shift/reduce conflicts, 3 reduce/reduce conflicts.
1077 conflicts
95 terminal symbols
92 non-terminal symbols
187 total symbols
394 rules
686 states

$>mybison  cfront3-naked.y
cfront3-naked.y: warning: 20 shift/reduce conflicts [-Wconflicts-sr]
cfront3-naked.y: warning: 4 reduce/reduce conflicts [-Wconflicts-rr]
cfront3-naked.y: note: rerun with option '-Wcounterexamples' to generate conflict counterexamples

$>mybison -z cfront3-naked.y # parse again but ignoring all precedences
cfront3-naked.y: warning: 1073 shift/reduce conflicts [-Wconflicts-sr]
cfront3-naked.y: warning: 4 reduce/reduce conflicts [-Wconflicts-rr]
cfront3-naked.y: note: rerun with option '-Wcounterexamples' to generate conflict counterexamples
cfront3-naked.y:251.11-29: warning: rule useless in parser due to conflicts [-Wother]
  251 | 	| e /*14:1*/ %prec GT
      |           ^~~~~~~~~~~~~~~~~~~
```

## Testing cql grammar (https://github.com/facebookincubator/CG-SQL/blob/main/sources/cql.y):
```
$>byacc-nb -E cql.y # convert yacc grammar to lemon grammar

$>mv y.yl cql.yl # rename converted lemon grammar

$>lemon-nb -s cql.yl
Parser statistics:
  terminal symbols...................   246
  non-terminal symbols...............   234
  total symbols......................   480
  rules..............................   695

byacc/CHANGES
2021-08-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* reader.c:
	fix memory-leak when replacing $$'s in destructor code (report/testcase
	by Boris Kolpackov).

	* main.c: account for a memory-leak

	* test/btyacc/btyacc_calc1.tab.c, test/btyacc/btyacc_demo.tab.c, test/btyacc/btyacc_destroy1.tab.c, test/btyacc/btyacc_destroy2.tab.c, test/btyacc/btyacc_destroy3.tab.c, test/btyacc/calc.tab.c, test/btyacc/calc1.tab.c, test/btyacc/calc2.tab.c, test/btyacc/calc3.tab.c, test/btyacc/calc_code_all.tab.c, test/btyacc/calc_code_default.tab.c, test/btyacc/calc_code_imports.tab.c, test/btyacc/calc_code_provides.tab.c, test/btyacc/calc_code_requires.tab.c, test/btyacc/calc_code_top.tab.c, test/btyacc/code_calc.code.c, test/btyacc/code_error.code.c, test/btyacc/defines1.calc.c, test/btyacc/defines2.calc.c, test/btyacc/defines3.calc.c, test/btyacc/empty.tab.c, test/btyacc/err_inherit3.tab.c, test/btyacc/err_inherit4.tab.c, test/btyacc/err_syntax10.tab.c, test/btyacc/err_syntax11.tab.c, test/btyacc/err_syntax12.tab.c, test/btyacc/err_syntax18.tab.c, test/btyacc/err_syntax20.tab.c, test/btyacc/error.tab.c, test/btyacc/expr.oxout.tab.c, test/btyacc/grammar.tab.c, test/btyacc/inherit0.tab.c, test/btyacc/inherit1.tab.c, test/btyacc/inherit2.tab.c, test/btyacc/ok_syntax1.tab.c, test/btyacc/pure_calc.tab.c, test/btyacc/pure_error.tab.c, test/btyacc/quote_calc-s.tab.c, test/btyacc/quote_calc.tab.c, test/btyacc/quote_calc2-s.tab.c, test/btyacc/quote_calc2.tab.c, test/btyacc/quote_calc3-s.tab.c, test/btyacc/quote_calc3.tab.c, test/btyacc/quote_calc4-s.tab.c, test/btyacc/quote_calc4.tab.c, test/btyacc/rename_debug.c, test/btyacc/stdin1.calc.c, test/btyacc/stdin2.calc.c, test/btyacc/varsyntax_calc1.tab.c, test/yacc/calc.tab.c, test/yacc/calc1.tab.c, test/yacc/calc2.tab.c, test/yacc/calc3.tab.c, test/yacc/calc_code_all.tab.c, test/yacc/calc_code_default.tab.c, test/yacc/calc_code_imports.tab.c, test/yacc/calc_code_provides.tab.c, test/yacc/calc_code_requires.tab.c, test/yacc/calc_code_top.tab.c, test/yacc/code_calc.code.c, test/yacc/code_error.code.c, test/yacc/defines1.calc.c, test/yacc/defines2.calc.c, test/yacc/defines3.calc.c, test/yacc/empty.tab.c, test/yacc/err_syntax10.tab.c, test/yacc/err_syntax11.tab.c, test/yacc/err_syntax12.tab.c, test/yacc/err_syntax18.tab.c, test/yacc/err_syntax20.tab.c, test/yacc/error.tab.c, test/yacc/expr.oxout.tab.c, test/yacc/grammar.tab.c, test/yacc/ok_syntax1.tab.c, test/yacc/pure_calc.tab.c, test/yacc/pure_error.tab.c, test/yacc/quote_calc-s.tab.c, test/yacc/quote_calc.tab.c, test/yacc/quote_calc2-s.tab.c, test/yacc/quote_calc2.tab.c, test/yacc/quote_calc3-s.tab.c, test/yacc/quote_calc3.tab.c, test/yacc/quote_calc4-s.tab.c, test/yacc/quote_calc4.tab.c, test/yacc/rename_debug.c, test/yacc/stdin1.calc.c, test/yacc/stdin2.calc.c, test/yacc/varsyntax_calc1.tab.c, btyaccpar.c, yaccpar.c:
	regen

	* btyaccpar.skel, yaccpar.skel:
	revert change to initialization of yystate, which confuses gcc, making a
	different warning

	* VERSION, package/byacc.spec, package/debian/changelog, package/mingw-byacc.spec, package/pkgsrc/Makefile:
	bump

2021-08-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* main.c, yacc.1: add "-h" option

	* test/btyacc/no_b_opt.error, test/btyacc/no_output2.error, test/btyacc/no_p_opt.error, test/btyacc/big_b.error, test/btyacc/big_l.error, test/btyacc/help.error, test/btyacc/nostdin.error, test/yacc/big_b.error, test/yacc/big_l.error, test/yacc/help.error, test/yacc/no_b_opt.error, test/yacc/no_output2.error, test/yacc/no_p_opt.error, test/yacc/nostdin.error:
	regen

	* main.c:
	map any of bison's long-options which have a corresponding yacc option
	into the latter, without depending upon getopt_long().

	* main.c: suggested patch:
	From: Boris Kolpackov <boris@codesynthesis.com>
	Subject: Re: [PATCH] support bison's --defines and --output options in byacc

	* VERSION, package/byacc.spec, package/debian/changelog, package/mingw-byacc.spec, package/pkgsrc/Makefile:
	bump


byacc/main.c
/* $Id: main.c,v 1.73 2021/08/08 20:39:34 tom Exp $ */

#include <signal.h>
#if !defined(_WIN32) || defined(__MINGW32__)
#include <unistd.h>		/* for _exit() */
#else
#include <stdlib.h>		/* for _exit() */
#endif

#include "defs.h"

#ifdef HAVE_MKSTEMP
# define USE_MKSTEMP 1
#elif defined(HAVE_FCNTL_H)
# define USE_MKSTEMP 1
# include <fcntl.h>		/* for open(), O_EXCL, etc. */
#else
# define USE_MKSTEMP 0
#endif

#if USE_MKSTEMP
#include <sys/types.h>
#include <sys/stat.h>

typedef struct _my_tmpfiles
{
    struct _my_tmpfiles *next;
    char *name;
}
MY_TMPFILES;

static MY_TMPFILES *my_tmpfiles;
#endif /* USE_MKSTEMP */

char dflag;
char dflag2;
char gflag;
char iflag;
char lflag;
static char oflag;
char rflag;
char sflag;
char tflag;
char vflag;
char ebnf_flag;
char lemon_flag;
char lemon_prec_flag;
char ignore_prec_flag;
char naked_flag;

const char *symbol_prefix;
const char *myname = "yacc";

int lineno;
int outline;

static char default_file_prefix[] = "y";

static char *file_prefix = default_file_prefix;

char *code_file_name;
char *input_file_name;
size_t input_file_name_len = 0;
char *defines_file_name;
char *externs_file_name;

static char *graph_file_name;
static char *output_file_name;
static char *verbose_file_name;
static char *ebnf_file_name;
static char *lemon_file_name;
static char *naked_file_name;

FILE *action_file;	/*  a temp file, used to save actions associated    */
			/*  with rules until the parser is written          */
FILE *code_file;	/*  y.code.c (used when the -r option is specified) */
FILE *defines_file;	/*  y.tab.h                                         */
FILE *externs_file;	/*  y.tab.i                                         */
FILE *input_file;	/*  the input file                                  */
FILE *output_file;	/*  y.tab.c                                         */
FILE *text_file;	/*  a temp file, used to save text until all        */
			/*  symbols have been defined                       */
FILE *union_file;	/*  a temp file, used to save the union             */
			/*  definition until all symbol have been           */
			/*  defined                                         */
FILE *verbose_file;	/*  y.output                                        */
FILE *ebnf_file;	/*  y.ebnf                                          */
FILE *lemon_file;	/*  lemon.y                                          */
FILE *naked_file;	/*  naked.y                                          */
FILE *graph_file;	/*  y.dot                                           */

Value_t nitems;
Value_t nrules;
Value_t nsyms;
Value_t ntokens;
Value_t nvars;

Value_t start_symbol;
char **symbol_name;
char **symbol_pname;
Value_t *symbol_value;
Value_t *symbol_prec;
char *symbol_assoc;

int pure_parser;
int token_table;
int error_verbose;

#if defined(YYBTYACC)
Value_t *symbol_pval;
char **symbol_destructor;
char **symbol_type_tag;
int locations = 0;	/* default to no position processing */
int backtrack = 0;	/* default is no backtracking */
char *initial_action = NULL;
#endif

int exit_code;

Value_t *ritem;
Value_t *rlhs;
Value_t *rrhs;
Value_t *rprec;
Assoc_t *rassoc;
bucket** rprec_bucket;
Value_t **derives;
char *nullable;

/*
 * Since fclose() is called via the signal handler, it might die.  Don't loop
 * if there is a problem closing a file.
 */
#define DO_CLOSE(fp) \
	if (fp != 0) { \
	    FILE *use = fp; \
	    fp = 0; \
	    fclose(use); \
	}

static int got_intr = 0;

void
done(int k)
{
    DO_CLOSE(input_file);
    DO_CLOSE(output_file);
    if (iflag)
	DO_CLOSE(externs_file);
    if (rflag)
	DO_CLOSE(code_file);

    DO_CLOSE(action_file);
    DO_CLOSE(defines_file);
    DO_CLOSE(graph_file);
    DO_CLOSE(text_file);
    DO_CLOSE(union_file);
    DO_CLOSE(verbose_file);

    if (got_intr)
	_exit(EXIT_FAILURE);

#ifdef NO_LEAKS
    DO_FREE(input_file_name);

    if (rflag)
	DO_FREE(code_file_name);

    if (dflag && !dflag2)
	DO_FREE(defines_file_name);

    if (iflag)
	DO_FREE(externs_file_name);

    if (oflag)
	DO_FREE(output_file_name);

    if (vflag)
	DO_FREE(verbose_file_name);

    if (ebnf_flag)
	DO_FREE(ebnf_file_name);

    if (lemon_flag)
	DO_FREE(lemon_file_name);

    if (naked_flag)
	DO_FREE(naked_file_name);

    if (gflag)
	DO_FREE(graph_file_name);

    lr0_leaks();
    lalr_leaks();
    mkpar_leaks();
    mstring_leaks();
    output_leaks();
    reader_leaks();
#endif

    exit(k);
}

static void
onintr(int sig GCC_UNUSED)
{
    got_intr = 1;
    done(EXIT_FAILURE);
}

static void
set_signals(void)
{
#ifdef SIGINT
    if (signal(SIGINT, SIG_IGN) != SIG_IGN)
	signal(SIGINT, onintr);
#endif
#ifdef SIGTERM
    if (signal(SIGTERM, SIG_IGN) != SIG_IGN)
	signal(SIGTERM, onintr);
#endif
#ifdef SIGHUP
    if (signal(SIGHUP, SIG_IGN) != SIG_IGN)
	signal(SIGHUP, onintr);
#endif
}

#define SIZEOF(v) (sizeof(v) / sizeof((v)[0]))

/*
 * Long options are provided only as a compatibility aid for scripters.
 */
/* *INDENT-OFF* */
static const struct {
    const char long_opt[16];
    const char yacc_arg;
    const char yacc_opt;
} long_opts[] = {
    { "defines",     1, 'H' },
    { "file-prefix", 1, 'b' },
    { "graph",       0, 'g' },
    { "ebnf",        0, 'e' },
    { "lemon",       0, 'E' },
    { "help",        0, 'h' },
    { "naked",       0, 'n' },
    { "name-prefix", 1, 'p' },
    { "no-lines",    0, 'l' },
    { "output",      1, 'o' },
    { "version",     0, 'V' }
};
/* *INDENT-ON* */

/*
 * Usage-message is designed for 80 columns, with some unknowns.  Account for
 * those in the maximum width so that the usage message uses no relocatable
 * pointers.
 */
#define USAGE_COLS (80 + sizeof(DEFINES_SUFFIX) + sizeof(OUTPUT_SUFFIX))

static void
usage(void)
{
    /* *INDENT-OFF* */
    static const char msg[][USAGE_COLS] =
    {
	{ "  -b file_prefix        set filename prefix (default \"y.\")" },
	{ "  -B                    create a backtracking parser" },
	{ "  -d                    write definitions (" DEFINES_SUFFIX ")" },
	{ "  -e                    write ebnf grammar" },
	{ "  -E                    write lemon grammar" },
	{ "  -h                    print this help-message" },
	{ "  -H defines_file       write definitions to defines_file" },
	{ "  -i                    write interface (y.tab.i)" },
	{ "  -g                    write a graphical description" },
	{ "  -l                    suppress #line directives" },
	{ "  -L                    enable position processing, e.g., \"%locations\"" },
	{ "  -n                    write naked grammar" },
	{ "  -o output_file        (default \"" OUTPUT_SUFFIX "\")" },
	{ "  -p symbol_prefix      set symbol prefix (default \"yy\")" },
	{ "  -P                    create a reentrant parser, e.g., \"%pure-parser\"" },
	{ "  -r                    produce separate code and table files (y.code.c)" },
	{ "  -s                    suppress #define's for quoted names in %token lines" },
	{ "  -t                    add debugging support" },
	{ "  -v                    write description (y.output)" },
	{ "  -V                    show version information and exit" },
	{ "  -u                    ignore precedences" },
	{ "  -z                    use leftmost token for rule precedence" },
    };
    /* *INDENT-ON* */
    unsigned n;

    fflush(stdout);
    fprintf(stderr, "Usage: %s [options] filename\n", myname);

    fprintf(stderr, "\nOptions:\n");
    for (n = 0; n < SIZEOF(msg); ++n)
    {
	fprintf(stderr, "%s\n", msg[n]);
    }

    fprintf(stderr, "\nLong options:\n");
    for (n = 0; n < SIZEOF(long_opts); ++n)
    {
	fprintf(stderr, "  --%-20s-%c\n",
		long_opts[n].long_opt,
		long_opts[n].yacc_opt);
    }

    exit(EXIT_FAILURE);
}

static void
invalid_option(const char *option)
{
    fprintf(stderr, "invalid option: %s\n", option);
    usage();
}

static void
setflag(int ch)
{
    switch (ch)
    {
    case 'B':
#if defined(YYBTYACC)
	backtrack = 1;
#else
	unsupported_flag_warning("-B", "reconfigure with --enable-btyacc");
#endif
	break;

    case 'd':
	dflag = 1;
	dflag2 = 0;
	break;

    case 'e':
	ebnf_flag = 1;
	break;

    case 'E':
	lemon_flag = 1;
	break;

    case 'g':
	gflag = 1;
	break;

    case 'i':
	iflag = 1;
	break;

    case 'l':
	lflag = 1;
	break;

    case 'L':
#if defined(YYBTYACC)
	locations = 1;
#else
	unsupported_flag_warning("-L", "reconfigure with --enable-btyacc");
#endif
	break;

    case 'n':
	naked_flag = 1;
	break;

    case 'P':
	pure_parser = 1;
	break;

    case 'r':
	rflag = 1;
	break;

    case 's':
	sflag = 1;
	break;

    case 't':
	tflag = 1;
	break;

    case 'v':
	vflag = 1;
	break;

    case 'V':
	printf("%s - %s\n", myname, VERSION);
	exit(EXIT_SUCCESS);

    case 'y':
	/* noop for bison compatibility. byacc is already designed to be posix
	 * yacc compatible. */
	break;

    case 'u':
        ignore_prec_flag = 1;
	break;

    case 'z':
        lemon_prec_flag = 1;
	break;

    default:
	usage();
    }
}

static void
getargs(int argc, char *argv[])
{
    int i;
#ifdef HAVE_GETOPT
    int ch;
#endif

    /*
     * Map bison's long-options into yacc short options.
     */
    for (i = 1; i < argc; ++i)
    {
	char *a = argv[i];

	if (!strncmp(a, "--", 2))
	{
	    char *eqls;
	    size_t lc;
	    size_t len;

	    if ((len = strlen(a)) == 2)
		break;

	    if ((eqls = strchr(a, '=')) != NULL)
	    {
		len = (size_t)(eqls - a);
		if (len == 0 || eqls[1] == '\0')
		    invalid_option(a);
	    }

	    for (lc = 0; lc < SIZEOF(long_opts); ++lc)
	    {
		if (!strncmp(long_opts[lc].long_opt, a + 2, len - 2))
		{
		    if (eqls != NULL && !long_opts[lc].yacc_arg)
			invalid_option(a);
		    *a++ = '-';
		    *a++ = long_opts[lc].yacc_opt;
		    *a = '\0';
		    if (eqls)
		    {
			while ((*a++ = *++eqls) != '\0') /* empty */ ;
		    }
		    break;
		}
	    }
	    if (!strncmp(a, "--", 2))
		invalid_option(a);
	}
    }

#ifdef HAVE_GETOPT
    if (argc > 0)
	myname = argv[0];

    while ((ch = getopt(argc, argv, "Bb:dEeghH:ilLno:Pp:rstVvyuz")) != -1)
    {
	switch (ch)
	{
	case 'b':
	    file_prefix = optarg;
	    break;
	case 'h':
	    usage();
	    break;
	case 'H':
	    dflag = dflag2 = 1;
	    defines_file_name = optarg;
	    break;
	case 'o':
	    output_file_name = optarg;
	    break;
	case 'p':
	    symbol_prefix = optarg;
	    break;
	default:
	    setflag(ch);
	    break;
	}
    }
    if ((i = optind) < argc)
    {
	/* getopt handles "--" specially, while we handle "-" specially */
	if (!strcmp(argv[i], "-"))
	{
	    if ((i + 1) < argc)
		usage();
	    input_file = stdin;
	    return;
	}
    }
#else
    char *s;
    int ch;

    if (argc > 0)
	myname = argv[0];

    for (i = 1; i < argc; ++i)
    {
	s = argv[i];
	if (*s != '-')
	    break;
	switch (ch = *++s)
	{
	case '\0':
	    input_file = stdin;
	    if (i + 1 < argc)
		usage();
	    return;

	case '-':
	    ++i;
	    goto no_more_options;

	case 'b':
	    if (*++s)
		file_prefix = s;
	    else if (++i < argc)
		file_prefix = argv[i];
	    else
		usage();
	    continue;

	case 'H':
	    dflag = dflag2 = 1;
	    if (*++s)
		defines_file_name = s;
	    else if (++i < argc)
		defines_file_name = argv[i];
	    else
		usage();
	    continue;

	case 'o':
	    if (*++s)
		output_file_name = s;
	    else if (++i < argc)
		output_file_name = argv[i];
	    else
		usage();
	    continue;

	case 'p':
	    if (*++s)
		symbol_prefix = s;
	    else if (++i < argc)
		symbol_prefix = argv[i];
	    else
		usage();
	    continue;

	default:
	    setflag(ch);
	    break;
	}

	for (;;)
	{
	    switch (ch = *++s)
	    {
	    case '\0':
		goto end_of_option;

	    default:
		setflag(ch);
		break;
	    }
	}
      end_of_option:;
    }

  no_more_options:

#endif /* HAVE_GETOPT */
    if (i + 1 != argc)
	usage();
    input_file_name_len = strlen(argv[i]);
    input_file_name = TMALLOC(char, input_file_name_len + 1);
    NO_SPACE(input_file_name);
    strcpy(input_file_name, argv[i]);
}

void *
allocate(size_t n)
{
    void *p;

    p = NULL;
    if (n)
    {
	p = CALLOC(1, n);
	NO_SPACE(p);
    }
    return (p);
}

#define CREATE_FILE_NAME(dest, suffix) \
	dest = alloc_file_name(len, suffix)

static char *
alloc_file_name(size_t len, const char *suffix)
{
    char *result = TMALLOC(char, len + strlen(suffix) + 1);
    if (result == 0)
	no_space();
    strcpy(result, file_prefix);
    strcpy(result + len, suffix);
    return result;
}

static char *
find_suffix(char *name, const char *suffix)
{
    size_t len = strlen(name);
    size_t slen = strlen(suffix);
    if (len >= slen)
    {
	name += len - slen;
	if (strcmp(name, suffix) == 0)
	    return name;
    }
    return NULL;
}

static void
create_file_names(void)
{
    size_t len;
    const char *defines_suffix;
    const char *externs_suffix;
    char *suffix;

    suffix = NULL;
    defines_suffix = DEFINES_SUFFIX;
    externs_suffix = EXTERNS_SUFFIX;

    /* compute the file_prefix from the user provided output_file_name */
    if (output_file_name != 0)
    {
	if (!(suffix = find_suffix(output_file_name, OUTPUT_SUFFIX))
	    && (suffix = find_suffix(output_file_name, ".c")))
	{
	    defines_suffix = ".h";
	    externs_suffix = ".i";
	}
    }

    if (suffix != NULL)
    {
	len = (size_t)(suffix - output_file_name);
	file_prefix = TMALLOC(char, len + 1);
	NO_SPACE(file_prefix);
	strncpy(file_prefix, output_file_name, len)[len] = 0;
    }
    else
	len = strlen(file_prefix);

    /* if "-o filename" was not given */
    if (output_file_name == 0)
    {
	oflag = 1;
	CREATE_FILE_NAME(output_file_name, OUTPUT_SUFFIX);
    }

    if (rflag)
    {
	CREATE_FILE_NAME(code_file_name, CODE_SUFFIX);
    }
    else
	code_file_name = output_file_name;

    if (dflag && !dflag2)
    {
	CREATE_FILE_NAME(defines_file_name, defines_suffix);
    }

    if (iflag)
    {
	CREATE_FILE_NAME(externs_file_name, externs_suffix);
    }

    if (vflag)
    {
	CREATE_FILE_NAME(verbose_file_name, VERBOSE_SUFFIX);
    }

    if (ebnf_flag)
    {
	CREATE_FILE_NAME(ebnf_file_name, EBNF_SUFFIX);
    }

    if (lemon_flag)
    {
	CREATE_FILE_NAME(lemon_file_name, LEMON_SUFFIX);
    }

    if (naked_flag)
    {
	CREATE_FILE_NAME(naked_file_name, NAKED_SUFFIX);
    }

    if (gflag)
    {
	CREATE_FILE_NAME(graph_file_name, GRAPH_SUFFIX);
    }

    if (suffix != NULL)
    {
	FREE(file_prefix);
    }
}

#if USE_MKSTEMP
static void
close_tmpfiles(void)
{
    while (my_tmpfiles != 0)
    {
	MY_TMPFILES *next = my_tmpfiles->next;

	(void)chmod(my_tmpfiles->name, 0644);
	(void)unlink(my_tmpfiles->name);

	free(my_tmpfiles->name);
	free(my_tmpfiles);

	my_tmpfiles = next;
    }
}

#ifndef HAVE_MKSTEMP
static int
my_mkstemp(char *temp)
{
    int fd;
    char *dname;
    char *fname;
    char *name;

    /*
     * Split-up to use tempnam, rather than tmpnam; the latter (like
     * mkstemp) is unusable on Windows.
     */
    if ((fname = strrchr(temp, '/')) != 0)
    {
	dname = strdup(temp);
	dname[++fname - temp] = '\0';
    }
    else
    {
	dname = 0;
	fname = temp;
    }
    if ((name = tempnam(dname, fname)) != 0)
    {
	fd = open(name, O_CREAT | O_EXCL | O_RDWR);
	strcpy(temp, name);
    }
    else
    {
	fd = -1;
    }

    if (dname != 0)
	free(dname);

    return fd;
}
#define mkstemp(s) my_mkstemp(s)
#endif

#endif

/*
 * tmpfile() should be adequate, except that it may require special privileges
 * to use, e.g., MinGW and Windows 7 where it tries to use the root directory.
 */
static FILE *
open_tmpfile(const char *label)
{
#define MY_FMT "%s/%.*sXXXXXX"
    FILE *result;
#if USE_MKSTEMP
    const char *tmpdir;
    char *name;

    if (((tmpdir = getenv("TMPDIR")) == 0 || access(tmpdir, W_OK) != 0) ||
	((tmpdir = getenv("TEMP")) == 0 || access(tmpdir, W_OK) != 0))
    {
#ifdef P_tmpdir
	tmpdir = P_tmpdir;
#else
	tmpdir = "/tmp";
#endif
	if (access(tmpdir, W_OK) != 0)
	    tmpdir = ".";
    }

    /* The size of the format is guaranteed to be longer than the result from
     * printing empty strings with it; this calculation accounts for the
     * string-lengths as well.
     */
    name = malloc(strlen(tmpdir) + sizeof(MY_FMT) + strlen(label));

    result = 0;
    if (name != 0)
    {
	int fd;
	const char *mark;

	mode_t save_umask = umask(0177);

	if ((mark = strrchr(label, '_')) == 0)
	    mark = label + strlen(label);

	sprintf(name, MY_FMT, tmpdir, (int)(mark - label), label);
	fd = mkstemp(name);
	if (fd >= 0
	    && (result = fdopen(fd, "w+")) != 0)
	{
	    MY_TMPFILES *item;

	    if (my_tmpfiles == 0)
	    {
		atexit(close_tmpfiles);
	    }

	    item = NEW(MY_TMPFILES);
	    NO_SPACE(item);

	    item->name = name;
	    NO_SPACE(item->name);

	    item->next = my_tmpfiles;
	    my_tmpfiles = item;
	}
	else
	{
	    FREE(name);
	}
	(void)umask(save_umask);
    }
#else
    result = tmpfile();
#endif

    if (result == 0)
	open_error(label);
    return result;
#undef MY_FMT
}

static void
open_files(void)
{
    create_file_names();

    if (input_file == 0)
    {
	input_file = fopen(input_file_name, "r");
	if (input_file == 0)
	    open_error(input_file_name);
    }

    action_file = open_tmpfile("action_file");
    text_file = open_tmpfile("text_file");

    if (vflag)
    {
	verbose_file = fopen(verbose_file_name, "w");
	if (verbose_file == 0)
	    open_error(verbose_file_name);
    }

    if (ebnf_flag)
    {
	ebnf_file = fopen(ebnf_file_name, "w");
	if (ebnf_file == 0)
	    open_error(ebnf_file_name);
    }

    if (lemon_flag)
    {
	lemon_file = fopen(lemon_file_name, "w");
	if (lemon_file == 0)
	    open_error(lemon_file_name);
    }

    if (naked_flag)
    {
	naked_file = fopen(naked_file_name, "w");
	if (naked_file == 0)
	    open_error(naked_file_name);
    }

    if (gflag)
    {
	graph_file = fopen(graph_file_name, "w");
	if (graph_file == 0)
	    open_error(graph_file_name);
	fprintf(graph_file, "digraph %s {\n", file_prefix);
	fprintf(graph_file, "\tedge [fontsize=10];\n");
	fprintf(graph_file, "\tnode [shape=box,fontsize=10];\n");
	fprintf(graph_file, "\torientation=landscape;\n");
	fprintf(graph_file, "\trankdir=LR;\n");
	fprintf(graph_file, "\t/*\n");
	fprintf(graph_file, "\tmargin=0.2;\n");
	fprintf(graph_file, "\tpage=\"8.27,11.69\"; // for A4 printing\n");
	fprintf(graph_file, "\tratio=auto;\n");
	fprintf(graph_file, "\t*/\n");
    }

    if (dflag || dflag2)
    {
	defines_file = fopen(defines_file_name, "w");
	if (defines_file == 0)
	    open_error(defines_file_name);
	union_file = open_tmpfile("union_file");
    }

    if (iflag)
    {
	externs_file = fopen(externs_file_name, "w");
	if (externs_file == 0)
	    open_error(externs_file_name);
    }

    output_file = fopen(output_file_name, "w");
    if (output_file == 0)
	open_error(output_file_name);

    if (rflag)
    {
	code_file = fopen(code_file_name, "w");
	if (code_file == 0)
	    open_error(code_file_name);
    }
    else
	code_file = output_file;
}

int
main(int argc, char *argv[])
{
    SRexpect = -1;
    RRexpect = -1;
    exit_code = EXIT_SUCCESS;

    set_signals();
    getargs(argc, argv);
    open_files();
    reader();
    lr0();
    lalr();
    make_parser();
    graph();
    finalize_closure();
    verbose();
    output();
    done(exit_code);
    /*NOTREACHED */
}
